### PR TITLE
[Bugfix] Chunk native host registration for large pinned pools

### PR DESCRIPTION
**Pull request by jooho-XCENA**
**What this PR does / why we need it**:

Large host KV-cache pools can fail to initialize on NVIDIA R580+ even when enough host memory is available. With 4 KiB pages, a single 512 GiB registration requires 2 GiB of driver page-table metadata, exceeding the `INT_MAX` limit of the driver's `kvzalloc()` path. This was reproduced with vLLM + LMCache on CXL-backed memory; anonymous mmap and tmpfs-backed shared memory hit the same per-call limit.

Register native mmap-backed pools in chunks of at most 64 GiB and unregister each chunk at its original address. This covers the NUMA, hugepage and shared-memory allocators. For `alloc_pinned_ptr`, requests of at least 512 GiB with Default/Portable/Mapped flags use anonymous mmap plus chunked registration; smaller requests retain `cudaHostAlloc`.

Registration failure rolls back earlier chunks in reverse order and clears the pending CUDA error through the same runtime. Free attempts every chunk even if one unregister reports an error. A mutex-protected ownership map lets the existing pointer-only `free_pinned_ptr` API select the matching release path, without changing callers.

References:
- [NVIDIA/open-gpu-kernel-modules#1234](https://github.com/NVIDIA/open-gpu-kernel-modules/pull/1234): proposed driver fix and our reproduction report, including successful registration of the same buffer in smaller chunks. The driver PR is still open.
- [vllm-project/vllm#51080](https://github.com/vllm-project/vllm/issues/51080): independent reproduction of the same 512 GiB per-call boundary during CPU KV offload on H200.
- [vllm-project/vllm#51081](https://github.com/vllm-project/vllm/pull/51081): KV-offload workaround using 64 GiB registrations, rollback and CUDA error draining; merged into vLLM main on September 17, 2026. This PR applies that size cap to LMCache's native allocator paths. vLLM's application-level workaround does not fix the underlying driver or LMCache's independent native allocation calls.

**Special notes for your reviewers**:

`cudaHostAllocWriteCombined` has no registration equivalent. Requests carrying it (or unknown flags) continue through `cudaHostAlloc`, preserving CUDA's semantics rather than silently dropping flags; the large-allocation workaround does not cover that case. The separate platform pin-memory API and lazy allocator are outside this change.

Validation:
- CPU-only regression harness compiles the production C++ allocator against a fake CUDA runtime. It checks the 512 GiB boundary, 520 GiB and partial tails, flag preservation, failure at each registration, reverse rollback, retry, unregister errors, and the NUMA/hugepage/shm allocation/free paths. Large mappings reserve virtual addresses only.
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python -m pytest --noconftest tests/v1/test_native_mem_alloc.py -q` — passed.
- Real CUDA 13.0 build and 4 MiB H2D/D2H round trips for ordinary pinned, shared-memory and NUMA allocations — passed on RTX PRO 6000 Blackwell, driver 580.126.20.
- `SKIP=rust-fmt,rust-clippy pre-commit run --all-files` — passed. Rust hooks explicitly skipped; no Rust changes.

The full integration suite and a physical allocation of at least 512 GiB were not rerun for this revision. The earlier local workaround recorded a successful 520 GiB allocation; this revision tests the large-size control flow with the fake runtime.

**If applicable**:

- [ ] this PR contains user facing changes - docs added
- [x] this PR contains unit tests

### CODE REVIEW

File: csrc/cuda/mem_alloc.h [hunk 1]
```diff
@@ -4,6 +4,11 @@
 #include <string>
 #include <vector>
 
+// Allocate size bytes of CUDA-pinned host memory; throws std::runtime_error on
+// allocation/registration failure. At >=512 GiB, Default/Portable/Mapped use
+// mmap with <=64 GiB registrations. Other flags retain cudaHostAlloc semantics.
+// Release only with free_pinned_ptr. Allocation bookkeeping may throw
+// bad_alloc.
```
Comment (deng451e) [inline csrc/cuda/mem_alloc.h:11]: This comment feels a bit too implementation specific for the public header. Could we remove it here?

File: csrc/cuda/mem_alloc.cpp [hunk 2]
```diff
@@ -18,6 +22,68 @@ static inline size_t _align_hugepage(size_t size) {
...
+        undo -= PIN_CHUNK_MAX;
+        cudaHostUnregister(base + undo);
+      }
+      _clear_cuda_error();
+      throw std::runtime_error(std::string("cudaHostRegister failed: ") +
+                               cudaGetErrorString(st));
+    }
+    off += n;
+  }
+}
```
Comment (deng451e) [inline csrc/cuda/mem_alloc.cpp:59]: Registration boundaries may split allocations. Fixed 64 GiB chunks can bisect 4 KiB-aligned objects, potentially breaking async copies or device-pointer access across independently registered ranges. Please align registration boundaries with allocation boundaries or make transfers chunk-aware, and add a real CUDA test that crosses a boundary.

File: tests/v1/native_mem_alloc/test_mem_alloc.cpp [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+      auto ptr = alloc_pinned_ptr(size, flags);
+      assert(alloc_size == size && alloc_flags == flags);
+      assert(registrations.empty());
+      free_pinned_ptr(ptr);
+      assert(free_calls == 1);
+    }
+  }
+  // Exact boundary, full chunks, and a tail that is not page-sized.
+  for (size_t size : {512 * GiB, 520 * GiB, 512 * GiB + 17}) {
+    for (unsigned int flags : {0u, 1u, 2u, 3u}) {
```
Comment (deng451e) [inline tests/v1/native_mem_alloc/test_mem_alloc.cpp:150]: The cutoff is one page too high. With 4 KiB pages, the largest safe single registration is 512 GiB - 4 KiB; a 512 GiB - 1 request still rounds up to enough pages to require 2 GiB of metadata. Please calculate the cutoff using the rounded page count

File: csrc/cuda/mem_alloc.cpp [hunk 4]
```diff
@@ -18,6 +22,68 @@ static inline size_t _align_hugepage(size_t size) {
...
+    cudaError_t st = cudaHostRegister(base + off, n, flags);
+    if (st != cudaSuccess) {
+      _clear_cuda_error();
+      for (size_t undo = off; undo > 0;) {
+        undo -= PIN_CHUNK_MAX;
+        cudaHostUnregister(base + undo);
+      }
+      _clear_cuda_error();
+      throw std::runtime_error(std::string("cudaHostRegister failed: ") +
+                               cudaGetErrorString(st));
```
Comment (deng451e) [inline csrc/cuda/mem_alloc.cpp:55]: Failed rollback may leave stale registrations. Rollback ignores cudaHostUnregister failures and then calls munmap, potentially leaving the driver referencing unmapped memory. Please handle each unregister failure and avoid unmapping memory that may still be registered.

**Comment by deng451e** [comment]
Hi, thanks for the contribution. my codex raised a few points that look valid to me. Could you please review the inline comments and address them?

**Comment by DongDongJu** [comment]
QQ, Is this for in-process mode? mp mode already doing 64mib pinning.

**Review by DongDongJu — APPROVED** [review]
Hello @jooho-XCENA,
IIUC, Its only applicable for eager allocation case(non-lazy) right?
Please address the comments from weishu. Otherwise LGTM.
